### PR TITLE
Packaging updates after templated RPC developments

### PR DESCRIPTION
**Pull request by jsturdy**
## Description
This PR adds changes for `gitlab` CI migration along with restructuring of the `xhalarm`/`xhalcore` into a single `xhal` package that builds libraries for different architectures (currently `arm` and `x86_64`)

Changes in the code are whitespace and formatting
* Updated all `#include` directives based on the new structure
* Added additional namespaces, depending on the subpackage (`common`, `client`, `server`)
* Made uniform 4 spaces indentation in all header files (not yet in `cpp` files though this can be added
* Made uniform the style in `cpp` files of not using `namespace blah {` scoping, in favour of `blah::` scoping
* Made uniform the style for `template` declarations:
  * `template<blah>` goes on it's own line
  * multi-line `template<blah,...>` always ends with `>` and the next line starts the next portion of the statement

### Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Motivation and Context
* The CI changes are introduced to facilitate migration to gitlab and automatic publication of releases.
* The restructuring was discussed in #132 and #131 and is seen as the natural progression of the migration to the templated framework

## How Has This Been Tested?
* No code changes were introduced, only formatting.
* The code compiles.
* All the intermediary CI related commits will be squashed before this is merge-ready, but for now, we can treat the code migration as reviewable.

### CODE REVIEW

**Comment by lmoureaux** [comment]
Where is this commit? It doesn't build without
```
[lmoureau@gem904daq01 test]$ git clone git@github.com:jsturdy/xhal.git -b citest/gitlab-ci-integration
Cloning into 'xhal'...
Enter passphrase for key '/afs/cern.ch/user/l/lmoureau/.ssh/id_rsa': 
remote: Enumerating objects: 160, done.
remote: Counting objects: 100% (160/160), done.
remote: Compressing objects: 100% (93/93), done.
remote: Total 7129 (delta 66), reused 151 (delta 62), pack-reused 6969
Receiving objects: 100% (7129/7129), 16.84 MiB | 9.22 MiB/s, done.
Resolving deltas: 100% (3994/3994), done.
Checking out files: 100% (4563/4563), done.
[lmoureau@gem904daq01 test]$ cd xhal/
[lmoureau@gem904daq01 xhal]$ git submodule update --init
Submodule 'config' (https://github.com/cms-gem-daq-project/gembuild.git) registered for path 'config'
Cloning into 'config'...
remote: Enumerating objects: 55, done.
remote: Counting objects: 100% (55/55), done.
remote: Compressing objects: 100% (41/41), done.
remote: Total 320 (delta 33), reused 29 (delta 14), pack-reused 265
Receiving objects: 100% (320/320), 108.34 KiB | 0 bytes/s, done.
Resolving deltas: 100% (201/201), done.
fatal: reference is not a tree: 38f4b5ac3f91bd04790e4f592a5e91ecac884b57
Unable to checkout '38f4b5ac3f91bd04790e4f592a5e91ecac884b57' in submodule path 'config'
```

**Comment by jsturdy** [comment]
It was cleaned up as part of the ongoing WIP cleaning.
This PR (as is) was not meant for testing the build, but rather for discussing the migrations/changes.
I can push a buildable version here, but I'm still ironing out some "features" of the gitlab CI, which have made the publishing step a little more tricky (jobs are done in a detached state so hard to pull the metadata I'm relying on for figuring out where to publish for all possible cases)

**Comment by lmoureaux** [comment]
It's really hard to understand what's going on with so many changes on various unrelated topics. Would dropping the CI commits be possible? Otherwise we should review them first, then do the migration (it's a bit unfortunate that these come at the same time).

**Comment by jsturdy** [comment]
Dropping the CI changes will have a negligible effect on this review... you can ignore anything in the `config` or `.gitlab-ci.yml` and basically be left with the migration related changes which would still be >95% of the changes.

I could have initially (~~but at this point am rather unwilling to go back and redo~~ might have to go back though, as it seems the `git log --follow` doesn't know about the moves for some reason, though I had checked this after the initial migration...) migrated all the code from old location to new locations and committed (knowing that it wouldn't compile) and then done all necessary fixup to get it to compile, and then done all the style alignment (though the difficulty of decoupling these last two was the reason everything was done in one go).

However, part of what I cleaned up recently was to move the changes around into three phases:
1) CI development that happened before your code merged
1) Refactoring the repo after your code merged
1) Additional tweaks to the CI after the refactor

These have now been force pushed

**Comment by jsturdy** [comment]
OK, I've started new such that history is preserved and have separated the various steps out:
1) migrating (easy part): done 
1) making things compilable (annoying part): ongoing
1) standardized style: following step
4) rebase CI developments on top (trivial): last step

I won't force push this until the last step is done

**Comment by lmoureaux** [comment]
```
arm-linux-gnueabihf-g++ ...
make[2]: arm-linux-gnueabihf-g++: Command not found
```
While we're at it, is there a way to detect this and display an error message? 

Here's my checklist. Note that it's what I found by manual inspection of the packages, I didn't try to install them:

- [x] Build works
- [x] Parallel build works
- [ ] `make rpm` works
```
/afs/cern.ch/work/l/lmoureau/GEM/test/xhal/config/ci/generate_repo.sh: line 43: EOS_SITE_URL: unbound variable
make[2]: *** [_rpmharvest] Error 1
make[2]: Leaving directory `/afs/cern.ch/work/l/lmoureau/GEM/test/xhal/python/reg_interface_gem'
make[1]: *** [reg_interface_gem.rpm] Error 2
make[1]: Leaving directory `/afs/cern.ch/work/l/lmoureau/GEM/test/xhal/python'
make: *** [python.rpm] Error 2
```
I set `EOS_SITE_URL=http://local.build/nowhere` for testing.

- [ ] Generated RPM contain stuff: Some of them are literally empty
```
$ ll $(find -name \*.rpm) -hS
-rw-r--r--. 1 lmoureau zh 1,3M 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/DEBUGRPMS/xhal/xhal-debuginfo-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh 283K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/DEBUGRPMS/xhal/xhal-debuginfo-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.armv7l.rpm
-rw-r--r--. 1 lmoureau zh 152K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-client-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh  36K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh  34K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/RPMS/xhal/ctp7-xhal-libs-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.noarch.rpm
-rw-r--r--. 1 lmoureau zh  32K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/RPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.armv7l.rpm
-rw-r--r--. 1 lmoureau zh  19K 23 sep 14:28 ./python/reg_interface_gem/rpm/repos/centos7_x86_64/RPMS/xhal/reg_interface_gem-1.0.0-1.0.338.dev.de4bf34git.centos7.python2.7.x86_64.rpm
-rw-r--r--. 1 lmoureau zh  16K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-devel-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh  12K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-server-devel-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh  10K 23 sep 14:28 ./python/reg_interface_gem/rpm/repos/SRPMS/xhal/reg_interface_gem-1.0.0-1.0.338.dev.de4bf34git.src.rpm
-rw-r--r--. 1 lmoureau zh 8,8K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-client-devel-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh 7,6K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/RPMS/xhal/xhal-server-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.armv7l.rpm
-rw-r--r--. 1 lmoureau zh 7,5K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-server-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
-rw-r--r--. 1 lmoureau zh 3,7K 23 sep 14:28 ./xhal/x86_64/rpm/repos/SRPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.src.rpm
-rw-r--r--. 1 lmoureau zh 3,6K 23 sep 14:28 ./xhal/arm/rpm/repos/SRPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.src.rpm
-rw-r--r--. 1 lmoureau zh 1,9K 23 sep 14:28 ./python/reg_interface_gem/rpm/repos/centos7_x86_64/DEBUGRPMS/xhal/reg_interface_gem-debuginfo-1.0.0-1.0.338.dev.de4bf34git.centos7.python2.7.x86_64.rpm
-rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./python/reg_interface_gem/rpm/reg_interface_gem.src.rpm
-rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./python/reg_interface_gem/rpm/reg_interface_gem.x86_64.rpm
-rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/arm/rpm/xhal.armv7l.rpm
-rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/arm/rpm/xhal.src.rpm
-rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/x86_64/rpm/xhal.src.rpm
-rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/x86_64/rpm/xhal.x86_64.rpm
```

- [ ] Dependencies are correct:
  * Shouldn't `-debug` depend on `-devel` for headers? Some debugged code will be located there...
  * `xhal-server` doesn't depend on `xhal`
  * `xhal-client` doesn't depend on `xhal`

- [ ] Packaging metadata is correct:
```
License     : __license__ <- Missing
Packager    : Mykhailo Dalchenko <- Should be CMS GEM Collaboration or Jared Sturdy
Summary     : None <- Should probably be filled as well
```
Empty metadata files:
```
-rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/doc/xhal-1.0.0/CHANGELOG.md
-rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/doc/xhal-1.0.0/MAINTAINER.md
-rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/doc/xhal-1.0.0/README.md
-rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/licenses/xhal-1.0.0/LICENSE
```

**Comment by jsturdy** [comment]
> ```
> arm-linux-gnueabihf-g++ ...
> make[2]: arm-linux-gnueabihf-g++: Command not found
> ```
> 
> While we're at it, is there a way to detect this and display an error message?
> 

Yeah, I can add a check if the compiler exists pretty easily, and print a message about the (likely) skipped setup

> Here's my checklist. Note that it's what I found by manual inspection of the packages, I didn't try to install them:
> 
>     * [x]  Build works
> 
>     * [x]  Parallel build works
> 
>     * [ ]  `make rpm` works
> 
> 
> ```
> /afs/cern.ch/work/l/lmoureau/GEM/test/xhal/config/ci/generate_repo.sh: line 43: EOS_SITE_URL: unbound variable
> make[2]: *** [_rpmharvest] Error 1
> make[2]: Leaving directory `/afs/cern.ch/work/l/lmoureau/GEM/test/xhal/python/reg_interface_gem'
> make[1]: *** [reg_interface_gem.rpm] Error 2
> make[1]: Leaving directory `/afs/cern.ch/work/l/lmoureau/GEM/test/xhal/python'
> make: *** [python.rpm] Error 2
> ```
> 
> I set `EOS_SITE_URL=http://local.build/nowhere` for testing.
> 

This was an oversight (as I hadn't been testing `make rpm` in my local area for some time), but now this dependency is moved to the actual publish step

>     * [ ]  Generated RPM contain stuff: Some of them are literally empty
> 
> 
> ```
> $ ll $(find -name \*.rpm) -hS
> -rw-r--r--. 1 lmoureau zh 1,3M 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/DEBUGRPMS/xhal/xhal-debuginfo-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh 283K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/DEBUGRPMS/xhal/xhal-debuginfo-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.armv7l.rpm
> -rw-r--r--. 1 lmoureau zh 152K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-client-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh  36K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh  34K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/RPMS/xhal/ctp7-xhal-libs-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.noarch.rpm
> -rw-r--r--. 1 lmoureau zh  32K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/RPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.armv7l.rpm
> -rw-r--r--. 1 lmoureau zh  19K 23 sep 14:28 ./python/reg_interface_gem/rpm/repos/centos7_x86_64/RPMS/xhal/reg_interface_gem-1.0.0-1.0.338.dev.de4bf34git.centos7.python2.7.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh  16K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-devel-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh  12K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-server-devel-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh  10K 23 sep 14:28 ./python/reg_interface_gem/rpm/repos/SRPMS/xhal/reg_interface_gem-1.0.0-1.0.338.dev.de4bf34git.src.rpm
> -rw-r--r--. 1 lmoureau zh 8,8K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-client-devel-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh 7,6K 23 sep 14:28 ./xhal/arm/rpm/repos/peta_armv7l/RPMS/xhal/xhal-server-1.0.0-1.0.338.dev.de4bf34git.peta.arm_linux_gnueabihf_gcc4_9_2.armv7l.rpm
> -rw-r--r--. 1 lmoureau zh 7,5K 23 sep 14:28 ./xhal/x86_64/rpm/repos/centos7_x86_64/RPMS/xhal/xhal-server-1.0.0-1.0.338.dev.de4bf34git.centos7.gcc4_8_5.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh 3,7K 23 sep 14:28 ./xhal/x86_64/rpm/repos/SRPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.src.rpm
> -rw-r--r--. 1 lmoureau zh 3,6K 23 sep 14:28 ./xhal/arm/rpm/repos/SRPMS/xhal/xhal-1.0.0-1.0.338.dev.de4bf34git.src.rpm
> -rw-r--r--. 1 lmoureau zh 1,9K 23 sep 14:28 ./python/reg_interface_gem/rpm/repos/centos7_x86_64/DEBUGRPMS/xhal/reg_interface_gem-debuginfo-1.0.0-1.0.338.dev.de4bf34git.centos7.python2.7.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./python/reg_interface_gem/rpm/reg_interface_gem.src.rpm
> -rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./python/reg_interface_gem/rpm/reg_interface_gem.x86_64.rpm
> -rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/arm/rpm/xhal.armv7l.rpm
> -rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/arm/rpm/xhal.src.rpm
> -rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/x86_64/rpm/xhal.src.rpm
> -rw-r--r--. 1 lmoureau zh    0 23 sep 14:28 ./xhal/x86_64/rpm/xhal.x86_64.rpm
> ```
> 

These empty ones are dummy files to help ensure that certain steps aren't reexecuted if they don't need to be (as a result of some `.PHONY` targets)

>     * [ ]  Dependencies are correct:
>       
>       * Shouldn't `-debug` depend on `-devel` for headers? Some debugged code will be located there...

Should be an easy change to the spec template

>       * `xhal-server` doesn't depend on `xhal`
>       * `xhal-client` doesn't depend on `xhal`
> 

Ah, good catch

>     * [ ]  Packaging metadata is correct:
> 
> 
> ```
> License     : __license__ <- Missing
> Packager    : Mykhailo Dalchenko <- Should be CMS GEM Collaboration or Jared Sturdy
> Summary     : None <- Should probably be filled as well
> ```
> 

Agreed, partially related to your next point

> Empty metadata files:
> 
> ```
> -rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/doc/xhal-1.0.0/CHANGELOG.md
> -rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/doc/xhal-1.0.0/MAINTAINER.md
> -rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/doc/xhal-1.0.0/README.md
> -rw-r--r--    1 root    root                        0 sep 23 14:36 /opt/xhal/share/licenses/xhal-1.0.0/LICENSE
> ```

These currently don't exist in the repo, so they are just dummies (well, the changelog will be generated automatically in the CI).
Need to make sure that the mandatory files are part of the repo, then they *should* be picked up

**Comment by jsturdy** [comment]
> > ```
> > arm-linux-gnueabihf-g++ ...
> > make[2]: arm-linux-gnueabihf-g++: Command not found
> > ```
> > 
> > 
> > While we're at it, is there a way to detect this and display an error message?
> 
> Yeah, I can add a check if the compiler exists pretty easily, and print a message about the (likely) skipped setup

Working on this (and trying to find a sensible solution, I think the best is actually to simply issue a warning about the missing compiler. My motivation for this is twofold:
1) not all targets need the compiler, so forgetting to set up the environment shouldn't really cause a failure
1) there is already an error when the compiler is missing, but what is missing is a helper about what should have been done, which can be added if initially the makefile detects that the compiler isn't present 

The only risk I see of this is that this warning *can* get lost in the output, since it won't necessarily be close to the error message.
An alternative is to run a `compiler-check` function as a prereq of any target that actually needs the compiler, but this to me is overengineering.

TL;DR I'm now leaning towards not addressing this at all, or to addressing it with an additional warning that a developer is supposed to scan the log to find.
Or to adding some additional info to the `make help` target

**Comment by mexanick** [comment]
were there any updated which didn't propagate here?

**Comment by jsturdy** [comment]
> were there any updated which didn't propagate here?

quite likely, i'll double check my working area and unset draft when ready

**Comment by jsturdy** [comment]
OK, @mexanick, @lpetre-ulb, @lmoureaux, I've pushed the latest tweaks I had been working on after cleanup

**Review by mexanick — APPROVED** [review]
Looks quite neat! Cursory approval. I think some parts (e.g. pythonic exposure example) will need to be updated, but since it is styling-only PR I'd leave it apart at the moment
